util/reward/reward.go

```diff
@@ -0,0 +1,24 @@
+package reward
+
+import (
+	"fmt"
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// CalculateRewardFactor returns the reward factor calculated using a logarmithic
+// model based on miss counters. missCount is the current miss count, m is the
+// maximum possible miss counts, and s is the smallest miss count in the period.
+// If the logarimthic function returns NaN the Reward Factor returned will be 0.
+// rewardFactor = 1 - logₘ₋ₛ₊₁(missCount - s + 1)
+func CalculateRewardFactor(missCount sdk.Dec, m sdk.Dec, s sdk.Dec) string {
+	logBase := m.Sub(s).Add(sdk.NewDec(1))
+	logKey := missCount.Sub(s).Add(sdk.NewDec(1))
+	rewardFactor := 1 - (math.Log(logKey.MustFloat64()) / math.Log(logBase.MustFloat64()))
```

Comment:
Are we able to use sdk.Dec all the way here? This still doesn't seem safe

Reply:
Do you know if the sdk has a math library with logarthmic functions? I didn't find any in their math package so I'm forced to use the std math library log functions which take and return floats.

Reply:
I don't think it contains any - as long as your operations perform in the same order every time, I think you're ok here. If there are any issues here we'll likely see them in determinism CI tests in the future, so I think you're okay for now